raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -632,7 +647,10 @@ public long epochStartOffset() {
     private ReplicaState getOrCreateReplicaState(ReplicaKey replicaKey) {
         ReplicaState state = voterStates.get(replicaKey.id());
         if (state == null || !state.matchesKey(replicaKey)) {
-            observerStates.putIfAbsent(replicaKey, new ReplicaState(replicaKey, false, Endpoints.empty()));
+            LeaderState.ReplicaState previous = observerStates.putIfAbsent(replicaKey, new ReplicaState(replicaKey, false, Endpoints.empty()));
+            if (previous == null) {
+                kafkaRaftMetrics.updateNumObservers(observerStates.size());
+            }
```

Comment:
Hmm. I am thinking that this optimization is not necessary at the moment. This applies to rest of the changes in this file.